test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java

```diff
@@ -0,0 +1,203 @@
+
+                if (i == NUM_ENTRIES -1) {
+                    // Help SparseOutputStream detect the last CEN entry write
+                    entry.setComment(LAST_COMMENT);
+                }
+                // Add the entry
+                zip.putNextEntry(entry);
+
+
+            }
+            // Finish writing the last entry
+            zip.closeEntry();
+
+            // Before the CEN headers are written, set the extra data on each entry
+            byte[] extra = makeLargeExtraField();
+            for (ZipEntry entry : entries) {
+                entry.setExtra(extra);
+            }
+        }
+    }
+
+    /**
+     * Validates that a ZipException is thrown with the expected message when
+     * the ZipFile is initialized with a ZIP whose CEN exeeds {@link #MAX_CEN_SIZE}
+     */
+    @Test
+    public void centralDirectoryTooLargeToFitInByteArray() {
+        ZipException ex = expectThrows(ZipException.class, () -> new ZipFile(hugeZipFile));
+        assertEquals(ex.getMessage(), "invalid END header (central directory size too large)");
+    }
+
+    /**
+     * We can reduce the number of written CEN headers by making each CEN header maximally large.
+     * We do this by adding the extra field produced by this method to each CEN header.
+     * <p>
+     * The structure of an extra field is as follows:
+     * <p>
+     * Header ID  (Two bytes, describes the type of the field, also called 'tag')
+     * Data Size  (Two byte short)
+     * Data Block (Contents depend on field type)
+     */
+    private byte[] makeLargeExtraField() {
+        // Make a maximally sized extra field
+        byte[] extra = new byte[MAX_EXTRA_FIELD_SIZE];
+        // Little-endian ByteBuffer for updating the header fields
+        ByteBuffer buffer = ByteBuffer.wrap(extra).order(ByteOrder.LITTLE_ENDIAN);
+
+        // We use the 'unknown' tag, specified in APPNOTE.TXT, 4.6.1 Third party mappings'
+        buffer.putShort(UNKNOWN_ZIP_TAG);
+
+        // Size of the actual (empty) data
+        buffer.putShort(MAX_DATA_SIZE);
+        return extra;
+    }
+
+    /**
+     * By writing sparse 'holes' until the last CEN is detected, we can save disk space
+     * used by this test from ~2GB to ~4K
+     */
+    private static class SparseOutputStream extends FilterOutputStream {
+        private final FileChannel channel;
+        private boolean afterLastCEN = false;
+        private long position = 0;
+
+        public SparseOutputStream(FileOutputStream fos) {
+            super(fos);
+            this.channel = fos.getChannel();
+        }
+
+        @Override
+        public void write(byte[] b, int off, int len) throws IOException {
+            position += len;
+            if (afterLastCEN) {
+                out.write(b, off, len);
+            } else {
+                channel.position(position);
+                if (Arrays.equals(LAST_COMMENT_BYTES, 0, LAST_COMMENT_BYTES.length, b, off, len)) {
```

Comment:
I would suggest a comment to help future developers.  I realize there are some earlier bread crumbs but I think this would be beneficial

Reply:
> I would suggest a comment to help future developers. I realize there are some earlier bread crumbs but I think this would be beneficial

I added a code comment explaining here, and also replaced the `afterLastCEN` variable with a `sparse` variable, flipping the if statements accordingly. Sparse is the 'normal' in this context, so better to have actual writes in the `else` branch.